bundles/org.eclipse.ui.ide.application/src/org/eclipse/ui/internal/ide/application/IDEWorkbenchAdvisor.java

```diff
@@ -539,31 +468,9 @@ protected void disconnectFromWorkspace() {
 		final MultiStatus status = new MultiStatus(IDEWorkbenchPlugin.IDE_WORKBENCH, 1,
 				IDEWorkbenchMessages.ProblemSavingWorkbench);
 		try {
-			final ProgressMonitorJobsDialog p = new CancelableProgressMonitorJobsDialog(
-					null);
-
-			final boolean applyPolicy = ResourcesPlugin.getWorkspace()
-					.getDescription().isApplyFileStatePolicy();
-
-			IRunnableWithProgress runnable = monitor -> {
-				try {
-					if (applyPolicy)
-						monitor = new CancelableProgressMonitorWrapper(monitor, p);
-
-					status.merge(((Workspace) ResourcesPlugin.getWorkspace()).save(true, true, monitor));
-				} catch (CoreException e) {
-					status.merge(e.getStatus());
-				}
-			};
+			status.merge(((Workspace) ResourcesPlugin.getWorkspace()).save(true, true, new NullProgressMonitor()));
 
-			p.run(true, false, runnable);
-		} catch (InvocationTargetException e) {
-			status
-					.merge(new Status(IStatus.ERROR,
-							IDEWorkbenchPlugin.IDE_WORKBENCH, 1,
-							IDEWorkbenchMessages.InternalError, e
-									.getTargetException()));
-		} catch (InterruptedException e) {
+		} catch (CoreException e) {
 			status.merge(new Status(IStatus.ERROR,
 					IDEWorkbenchPlugin.IDE_WORKBENCH, 1,
 					IDEWorkbenchMessages.InternalError, e));
```

Comment:
As mentioned [here](https://github.com/eclipse-platform/eclipse.platform.ui/issues/1269#issuecomment-1801276486), I would propose to execute the save operation in a `busyCursorWhile` of the `ProgressManager` , so that the UI does not freeze in case saving takes long. Instead a busy cursor will be shown first and when it takes a "long time" (currently more than 800ms), it will open a progress dialog.

```java
ProgressManager.getInstance().busyCursorWhile(monitor -> {
	try {
		status.merge(((Workspace) ResourcesPlugin.getWorkspace()).save(true, true, monitor));
	} catch (CoreException e) {
		status.merge(new Status(IStatus.ERROR,
					IDEWorkbenchPlugin.IDE_WORKBENCH, 1,
					IDEWorkbenchMessages.InternalError, e));
	}
});
```